File: 02_OOP/src/_01-introduction/control_flow/comparison/examples.py
# Instrukcje sterujące w Pythonie
# Porównaj z: Java (examples/) i C (examples.c)
#
# Uruchomienie: python examples.py

# ===== if / elif / else =====
# Python: elif (nie else if)
# Python: wcięcia zamiast nawiasów klamrowych!
temperature = 25

if temperature > 30:
    print("Upał")
elif temperature > 20:
    print(f"Ciepło: {temperature}°C")
else:
    print("Chłodno")

# ===== Operator ternary (wyrażenie warunkowe) =====
# Python: wartość_tak if warunek else wartość_nie (odwrócona kolejność!)
age = 20
status = "dorosły" if age >= 18 else "małoletni"
print(f"Wiek {age} → {status}")

# ===== match/case (Python 3.10+) — odpowiednik switch =====
# Python: match..case zamiast switch
day = 3
match day:
    case 1: print("Poniedziałek")
    case 2: print("Wtorek")
    case 3: print("Środa")
    case _: print("Inny dzień")  # _ to wildcard (default)

# ===== for — iteracja po zakresie =====
# Python: range(n) zamiast int i=0; i<n; i++
for i in range(5):
    print(i, end=" ")
print()

# for po liście (jak Java for-each)
fruits = ["Jabłko", "Banan", "Wiśnia"]
for fruit in fruits:
    print(f"  - {fruit}")

# enumerate — indeks + wartość
for idx, fruit in enumerate(fruits):
    print(f"  [{idx}] {fruit}")

# ===== while =====
n = 1
while n <= 1024:
    print(n, end=" ")
    n *= 2
print()

# ===== Brak do-while w Python! =====
# Ekwiwalent:
x = 10
while True:
    print(f"x={x}")
    x += 1
    if x >= 5:
        break

# ===== break / continue =====
for i in range(10):
    if i == 5:
        break
    if i % 2 == 0:
        continue
    print(i, end=" ")
print()

# ===== else w pętli (unikalna cecha Pythona!) =====
# else wykonuje się gdy pętla zakończyła się normalnie (bez break)
for i in range(5):
    print(i, end=" ")
else:
    print("\nPętla zakończona normalnie (bez break)")

# ===== Comprehensions (brak w Java/C) =====
squares = [x**2 for x in range(10)]
print("Kwadraty:", squares)

even = [x for x in range(20) if x % 2 == 0]
print("Parzyste:", even)

# RÓŻNICE Python vs Java:
#
# 1. Python: wcięcia definiują bloki (nie {})
#    Java: nawiasy klamrowe {}
#
# 2. Python: range() w for
#    Java: int i=0; i<n; i++
#
# 3. Python: brak do-while (emulacja przez while True + break)
#    Java: do { } while(...)
#
# 4. Python: match/case (Python 3.10) z pattern matching
#    Java: switch expression (Java 14) + pattern matching (Java 21)
#
# 5. Python: else po pętli
#    Java: brak odpowiednika
#
# 6. Python: list comprehensions [x for x in ...]
#    Java: Stream API (stream().filter().map().collect())
#
# 7. Python: brak goto (jak Java)
#    Java: labeled break/continue
#
# 8. Python: dynamiczne typowanie (brak deklaracji typu)
#    Java: statyczne typowanie (int i = 0)

File: 02_OOP/src/_01-introduction/control_flow/comparison/examples.c
/*
 * Instrukcje sterujące w języku C
 * Porównaj z: Java (examples/) i Python (examples.py)
 *
 * Kompilacja: gcc -o examples examples.c
 * Uruchomienie: ./examples
 */
#include <stdio.h>
#include <stdbool.h>

int main() {

    /* ===== if / else ===== */
    int temperature = 25;
    if (temperature > 30) {
        printf("Upał\n");
    } else if (temperature > 20) {
        printf("Ciepło: %d°C\n", temperature);
    } else {
        printf("Chłodno\n");
    }

    /* ===== Operator ternary ===== */
    int age = 20;
    const char* status = (age >= 18) ? "dorosły" : "małoletni";
    printf("Wiek %d → %s\n", age, status);

    /* ===== switch ===== */
    /* UWAGA: w C switch działa tylko dla int/char/enum */
    /* BRAK switch expression jak w Java 14+! */
    int day = 3;
    switch (day) {
        case 1: printf("Poniedziałek\n"); break;
        case 2: printf("Wtorek\n");       break;
        case 3: printf("Środa\n");        break;
        default: printf("Inny dzień\n");
    }

    /* ===== for ===== */
    /* Identyczne jak Java! (Java przejęła składnię z C) */
    for (int i = 0; i < 5; i++) {
        printf("%d ", i);
    }
    printf("\n");

    /* ===== while ===== */
    int n = 1;
    while (n <= 1024) {
        printf("%d ", n);
        n *= 2;
    }
    printf("\n");

    /* ===== do-while ===== */
    int x = 10;
    do {
        printf("x=%d\n", x);
        x++;
    } while (x < 5);

    /* ===== break / continue ===== */
    for (int i = 0; i < 10; i++) {
        if (i == 5) break;
        if (i % 2 == 0) continue;
        printf("%d ", i);
    }
    printf("\n");

    /* ===== goto — istnieje w C, BRAK w Java i Python! ===== */
    /* (nie zalecane, ale istnieje) */
    int counter = 0;
    loop:
        counter++;
        if (counter < 5) goto loop;
    printf("goto counter = %d\n", counter);

    return 0;
}

/*
 * RÓŻNICE C vs Java:
 *
 * 1. C: brak boolean (używa int 0/1 lub stdbool.h)
 *    Java: boolean jako typ prymitywny
 *
 * 2. C: switch tylko dla typów całkowitych (int, char, enum)
 *    Java: switch dla int, char, String, enum + pattern matching (Java 21)
 *
 * 3. C: brak for-each (brak iteratorów wbudowanych)
 *    Java: for-each dla tablic i kolekcji
 *
 * 4. C: goto istnieje (nie zalecane)
 *    Java: labeled break/continue zamiast goto
 *
 * 5. C: wskaźniki — arytmetyka adresów
 *    Java: referencje — brak arytmetyki
 *
 * 6. C: ręczne zarządzanie pamięcią (malloc/free)
 *    Java: automatyczny GC
 */

